Ingress and Terminating Gateway docs (#7710)
hashicorp/consul: website/pages/docs/connect/terminating_gateway.mdx

---
layout: docs
page_title: Connect - Terminating Gateways
sidebar_title: Terminating Gateways
description: >-
  A terminating gateway enables traffic from services in the Consul
  service mesh to services outside the mesh. This section details
  how to configure and run a terminating gateway.
---

# Terminating Gateways

-> **1.8.0+:**  This feature is available in Consul versions 1.8.0 and newer.

Terminating gateways enable connections from services in the Consul service mesh to
services outside the mesh. These gateways effectively act as Connect proxies that can
represent more than one service. They terminate Connect mTLS connections, enforce intentions,
and forward requests to the appropriate destination.

## Security Considerations

> We recommend that terminating gateways not be exposed to the WAN or open internet. This is because terminating gateways

Reviewer: ```suggestion
We recommend that terminating gateways are not exposed to the WAN or open internet. This is because terminating gateways
```

> We recommend that terminating gateways not be exposed to the WAN or open internet. This is because terminating gateways

Reviewer: You might consider moving this to the intro and putting it in a call out box so users don't miss it.

Reviewer 2: Moved it into a call out box but left it at the top of this section. Let me know what you think.

hold certificates to decrypt Consul Connect traffic directed at them and may be configured with credentials to connect
to linked services. Connections over the WAN or open internet should flow through [mesh gateways](/docs/connect/mesh_gateway)
whenever possible since they are not capable of decrypting traffic or connecting directly to services.

By specifying a path to a [CA file](/docs/agent/config-entries/terminating-gateway#cafile) connections

> from the gateway will be encrypted using one-way TLS authentication. If a path to a

Reviewer: ```suggestion
from the terminating gateway will be encrypted using one-way TLS authentication. If a path to a
```

[client certificate](/docs/agent/config-entries/terminating-gateway#certfile)
and [private key](/docs/agent/config-entries/terminating-gateway#keyfile) are also specified connections

> from the gateway will be encrypted using mutual TLS authentication.

Reviewer: ```suggestion
from the terminating gateway will be encrypted using mutual TLS authentication.
```

If none of these are provided, Consul will **only** encrypt connections to the gateway and not
from the gateway to the destination service.

> -> **Note:** If certificates and keys are configured the gateway will upgrade HTTP connections to TLS. 

Reviewer: ```suggestion
-> **Note:** If certificates and keys are configured the terminating gateway will upgrade HTTP connections to TLS. 
```

Client applications can issue plain HTTP requests even when connecting to servers that require HTTPS.

## Prerequisites

Each terminating gateway needs:

> 1. A local Consul agent to manage its configuration.

Reviewer: Local Consul agent sounds like a dev agent. Can it be updated to Consul client?

2. General network connectivity to services within its local Consul datacenter.

Terminating gateways also require that your Consul datacenters are configured correctly:

- You'll need to use Consul version 1.8.0.
- Consul [Connect](/docs/agent/options#connect) must be enabled on the datacenter's Consul servers.
- [gRPC](/docs/agent/options#grpc_port) must be enabled on all client agents.

> - [Centralized configuration](/docs/agent/options#enable_central_service_config) must be enabled.

Reviewer: I don't think this was a prereq in the docs for ingress gateways, should it be?

Reviewer 2: It doesn't need to be actually. Fixed.

Currently, [Envoy](https://www.envoyproxy.io/) is the only proxy with terminating gateway capabilities in Consul.

- Terminating gateway proxies receive their configuration through Consul, which
  automatically generates it based on the gateway's registration. Currently Consul
  can only translate terminating gateway registration information into Envoy
  configuration, therefore the proxies acting as terminating gateways must be Envoy.

> Connect proxies that send upstream traffic through a gateway aren't

Reviewer: Can we say "sidecar proxies" if referring to proxies for services?

Reviewer 2: I don't think so. This could also refer to Ingress gateways and 3rd party non-sidecar proxies that users might have.

Reviewer: I guess that's fine. We should note that somewhere though, maybe on the Connect overview page.

affected when you deploy terminating gateways. If you are using non-Envoy proxies as
Connect proxies they will continue to work for traffic directed at services linked to
a terminating gateway as long as they discover upstreams with the
[/health/connect](/api/health#list-nodes-for-connect-capable-service) endpoint.

## Running and Using a Terminating Gateway

You must complete the following steps to configure a terminating gateway to proxy traffic from services in the Consul service mesh:

> 1. On a dedicated host for the gateway, start an Envoy proxy using the [envoy subcommand](/docs/commands/connect/envoy#terminating-gateways) and

Reviewer: See note on ingress gateway docs.

Reviewer 2: Changed the wording a bit, let me know what you think.

specifying the `terminating` gateway type:
    ```shell
    $ consul connect envoy -gateway=terminating -register -service us-west-gateway \
      -address '{{ GetInterfaceIP "eth0" }}:8443'
    ```

2. Create and apply a `terminating-gateway` [configuration entry](/docs/agent/config-entries/terminating-gateway) that defines 
a set of services that the gateway will proxy traffic to. The config entry can be applied via the 
[CLI](/docs/commands/config/write) or [API](/api/config#apply-configuration).

> 3. Ensure that [Consul intentions](/docs/commands/intention) are set up between the source services in the mesh 

Reviewer: See notes on ingress, can we specify the direction of the intention?

4. Ensure necessary [upstreams](/docs/connect/registration/service-registration#upstream-configuration-reference)
have been added to service definitions of the source services in the Consul service mesh. Each source service that needs
to connect to a service proxied by a terminating gateway will need to add the external service as an upstream destination.

## Terminating Gateway Configuration

Terminating gateways are defined similarly to other services registered with Consul, with two exceptions.
The first is that the [kind](/api/agent/service#kind) must be "terminating-gateway". Second,
the terminating gateway service definition may contain a `Proxy.Config` entry just like a
Connect proxy service, to define opaque configuration parameters useful for the actual proxy software.
For Envoy there are some supported [gateway options](/docs/connect/proxies/envoy#gateway-options) as well as
[escape-hatch overrides](/docs/connect/proxies/envoy#escape-hatch-overrides).

-> **Note:** If ACLs are enabled, the terminating gateways must be registered with a token granting `service:write`

> for the gateway's service name **and** all linked services. These permissions will authorize the gateway

Reviewer: ```suggestion
for the gateway's service name **and** all linked services. These privileges will authorize the gateway
```

---
layout: docs
page_title: Connect - Terminating Gateways
sidebar_title: Terminating Gateways
description: >-
  A terminating gateway enables traffic from services in the Consul
  service mesh to services outside the mesh. This section details
  how to configure and run a terminating gateway.
---

# Terminating Gateways

-> **1.8.0+:**  This feature is available in Consul versions 1.8.0 and newer.

Terminating gateways enable connections from services in the Consul service mesh to
services outside the mesh. These gateways effectively act as Connect proxies that can
represent more than one service. They terminate Connect mTLS connections, enforce intentions,
and forward requests to the appropriate destination.

![Terminating Gateway Architecture](/img/terminating-gateways.png)

~> **Beta limitations:** Terminating Gateways currently do not support targeting service subsets with
[L7 configuration](/docs/connect/l7-traffic-management). They route to all instances of a service with no capabilities
for filtering by instance. Terminating Gateways also currently do not support routing to services with a hostname

> defined as a their address. The address registered with Consul for services the gateway will route to **must** be a resolved IP address.

Reviewer: ```suggestion
defined as a their address. The service address registered with Consul, that the gateway will route traffic to,  **must** be a resolved IP address.
```

- Consul [Connect](/docs/agent/options#connect) must be enabled on the datacenter's Consul servers.
- [gRPC](/docs/agent/options#grpc_port) must be enabled on all client agents.

Currently, [Envoy](https://www.envoyproxy.io/) is the only proxy with terminating gateway capabilities in Consul.

- Terminating gateway proxies receive their configuration through Consul, which
  automatically generates it based on the gateway's registration. Currently Consul
  can only translate terminating gateway registration information into Envoy
  configuration, therefore the proxies acting as terminating gateways must be Envoy.

Connect proxies that send upstream traffic through a gateway aren't
affected when you deploy terminating gateways. If you are using non-Envoy proxies as
Connect proxies they will continue to work for traffic directed at services linked to
a terminating gateway as long as they discover upstreams with the
[/health/connect](/api/health#list-nodes-for-connect-capable-service) endpoint.

## Running and Using a Terminating Gateway

You must complete the following steps to configure a terminating gateway to proxy traffic from services in the Consul service mesh:

> 1. On a host dedicated to the gateway and Consul client, start an Envoy proxy using the [envoy subcommand](/docs/commands/connect/envoy#terminating-gateways) and

Reviewer: ```suggestion
1. On a Consul client, that is also a dedicated host for the gateway, start an Envoy proxy using the [envoy subcommand](/docs/commands/connect/envoy#terminating-gateways) and
```

Reviewer 2: I think this doesn't really work, since a Consul client is an agent and not a host.

Rewording as "On a host with a Consul client agent" to also be less opinionated about it needed to be dedicated to the gateway.

Terminating gateways are defined in service configuration files and registered with Consul like other services, with two exceptions.
The first is that the [kind](/api/agent/service#kind) must be "terminating-gateway". Second,
the terminating gateway service definition may contain a `Proxy.Config` entry just like a
Connect proxy service, to define opaque configuration parameters useful for the actual proxy software.
For Envoy there are some supported [gateway options](/docs/connect/proxies/envoy#gateway-options) as well as
[escape-hatch overrides](/docs/connect/proxies/envoy#escape-hatch-overrides).

-> **Note:** If ACLs are enabled, the terminating gateways must be registered with a token granting `service:write`
for the gateway's service name **and** all linked services. These privileges will authorize the gateway
to terminate mTLS connections on behalf of the linked services.

Linking services to a terminating gateway is done with a `terminating-gateway`
[configuration entry](/docs/agent/config-entries/terminating-gateway). This config entry can be applied via the 
[CLI](/docs/commands/config/write) or [API](/api/config#apply-configuration).

> Gateways with the same name in Consul's service catalog are configured with a single configuration entry.

Reviewer: Terminating specifically or any gateway?

Reviewer 2: any gateway

This means that additional gateway instances registered with the same name will determine their routing based on the existing configuration entry.
Adding replicas of a gateway that routes to a particular set of services requires running the
[envoy subcommand](/docs/commands/connect/envoy#terminating-gateways) on additional hosts and specifying
the same gateway name with the `service` flag.

> ~> [Configuration Entries](/docs/agent/config-entries) are global in scope. A configuration entry for a gateway name applies

Reviewer: ```suggestion
~> [Configuration entries](/docs/agent/config-entries) are global in scope. A configuration entry for a gateway name applies
```

   across all federated Consul datacenters. If terminating gateways in different Consul datacenters need to route to different
   sets of services within their datacenter then the terminating gateways **must** be registered with different names.

> The services that the terminating gateway will proxy for must be registered with Consul. They must also be registered

Reviewer: ```suggestion
The services that the terminating gateway will proxy for must be registered with Consul, even the services outside the mesh. They must also be registered
```

in the same Consul datacenter as the terminating gateway. Otherwise the terminating gateway will not be able to
discover the services' addresses. These services can be registered with a local Consul agent.
If there is no agent present, the services can be registered [directly in the catalog](/api/catalog#register-entity)
by sending the registration request to a client or server agent on a different host.

> Since Consul's services must be associated with a node, direct registrations to the catalog must define a node as well.

Reviewer: I'm having a little trouble trying to understand the underlying issue. Is it that Consul registers all agents in the catalog by the node name and therefor all node names must be unique? Is there something else that TG introduces to make the issue more complicated?

Reviewer 2: Yes node names are unique. The main issue is that agents periodically query the catalog for services registered to their node name. If someone registers an external service to that same node name:
1. The agent will find out
2. It will see it has no record of that service locally
3. It will send a request to delete that service from the catalog so that the catalog reflects its local state

Reviewer: "All services registered in the Consul catalog must be associated with a node, include services registered directly with tools like Consul ESM. Ensure that services registered directly are associated with a node that has a unique name. Consul's anit-entropy will (then summarize those steps)"

---
layout: docs
page_title: Connect - Terminating Gateways
sidebar_title: Terminating Gateways
description: >-

>   A Terminating Gateway enables traffic from services in the Consul

Reviewer: ```suggestion
  A terminating gateway enables traffic from services in the Consul
```

Terminating gateways enable connections from services in the Consul service mesh to
services outside the mesh. These gateways effectively act as Connect proxies that can
represent more than one service. They terminate Connect mTLS connections, enforce intentions,
and forward requests to the appropriate destination.

For data that must be encrypted in transit to the final destination, terminating gateways
support TLS origination. By specifying a [CA file](/docs/agent/config-entries/terminating-gateway#cafile),
[client certificate](/docs/agent/config-entries/terminating-gateway#certfile), and
[private key](/docs/agent/config-entries/terminating-gateway#keyfile),
connections from the gateway to a linked service can be encrypted using one-way or mutual TLS.
If none of these are provided, Consul will only encrypt connections to the gateway and not
from the gateway to the destination service. 

-> **Note:** If certificates and keys are configured the gateway will upgrade HTTP connections to TLS. 
Client applications can issue plain HTTP requests even when connecting to servers that require HTTPS.

## Prerequisites

> Each terminating gateway needs two things:

Reviewer: ```suggestion
Each terminating gateway needs:
```

Reviewer: Do we also want to say it needs a supported proxy technology?

Reviewer 2: That's mentioned a little farther down:
https://github.com/hashicorp/consul/pull/7710/files#diff-bcc0b2977dadf67996a8a8ce825314a8R45

1. A local Consul agent to manage its configuration.
2. General network connectivity to services within its local Consul datacenter.

Terminating gateways also require that your Consul datacenters are configured correctly:

- You'll need to use Consul version 1.8.0.
- Consul [Connect](/docs/agent/options#connect) must be enabled on the datacenter's Consul servers.
- [gRPC](/docs/agent/options#grpc_port) must be enabled on all client agents.
- [Centralized configuration](/docs/agent/options#enable_central_service_config) must be enabled.

Currently, [Envoy](https://www.envoyproxy.io/) is the only proxy with terminating gateway capabilities in Consul.

- Terminating gateway proxies receive their configuration through Consul, which
  automatically generates it based on the gateway's registration. Currently Consul
  can only translate terminating gateway registration information into Envoy
  configuration, therefore the proxies acting as terminating gateways must be Envoy.

> Sidecar proxies that send upstream traffic through a gateway aren't

Reviewer: I prefer saying sidecar proxy when referring to the service proxy, however, we've been using the term Consul proxy or Connect proxy service. Maybe we can do a find/replace for both these docs so it's consistent?

affected when you deploy terminating gateways. If you are using non-Envoy proxies as
Connect sidecars they will continue to work for traffic directed at services linked to
a terminating gateway as long as they discover upstreams with the
[/health/connect](/api/health#list-nodes-for-connect-capable-service) endpoint.

## Running and Using a Terminating Gateway

You must complete the following steps to configure a terminating gateway to proxy traffic from services in the Consul service mesh:

1. Start an Envoy proxy using the [envoy subcommand](/docs/commands/connect/envoy#terminating-gateways), specifying the `terminating` gateway type.
2. Create a `terminating-gateway` [configuration entry](/docs/agent/config-entries/terminating-gateway) that defines 
a set of services that the gateway will proxy traffic to. The config entry can be applied via the 
[CLI](/docs/commands/config/write) or [API](/api/config#apply-configuration).

> 3. Ensure that [Consul intentions](/docs/commands/intention#consul-intention) are set up between the source services in the mesh 

Reviewer: ```suggestion
3. Ensure that [Consul intentions](/docs/commands/intention#consul-intention) are set up between the source services in the mesh. 
```

and the gateway's linked services. The intentions must reference the destination service's name and not the gateway itself.

## Terminating Gateway Configuration

> Ingress gateways are defined similarly to other services registered with Consul, with two exceptions.

Reviewer: ```suggestion
Terminating gateways are defined similarly to other services registered with Consul, with two exceptions.
```

The first is that the [kind](/api/agent/service#kind) must be "ingress-gateway". Second,
the ingress gateway service definition may contain a `Proxy.Config` entry just like a
Connect proxy service, to define opaque configuration parameters useful for the actual proxy software.

> Terminating gateways are defined similarly to other services registered with Consul, with two exceptions.

Reviewer: Is this and the above paragraph duplicates?

Reviewer 2: Woops, yes. Fixed.

The first is that the [kind](/api/agent/service#kind) must be "terminating-gateway". Second,
the terminating gateway service definition may contain a `Proxy.Config` entry just like a
Connect proxy service, to define opaque configuration parameters useful for the actual proxy software.
For Envoy there are some supported [gateway options](/docs/connect/proxies/envoy#gateway-options) as well as
[escape-hatch overrides](/docs/connect/proxies/envoy#escape-hatch-overrides).

> -> **Note:** If ACLs are enabled, the gateways must be registered with a token granting `service:write`

Reviewer: ```suggestion
-> **Note:** If ACLs are enabled, the terminating gateways must be registered with a token granting `service:write`
```

Below, I see several more plain gateways that should be updated.
